kf_lib_data_ingest/validation/data_validator.py

```diff
@@ -0,0 +1,352 @@
+                if bad_values:
+                    errors[fname] = bad_values
+
+        yield _format_result("attribute", f"{type} {desc}", valid, errors)
+
+
+def _build_graph(dict_of_dataframes, include_implicit=True):
+    """Construct a graph that represents the study data. Nodes are cells in the
+    data, edges are relationships between cells across rows according to the
+    designated relationship hierarchy.
+
+    :param dict_of_dataframes: dict with filename keys and dataframe values
+    :param include_implicit: whether to deduce implied connections, default True
+    :return: a graph representation of the data
+    """
+    prep = "WITH" if include_implicit else "WITHOUT"
+    logger.info(f"Building node graph {prep} implied connection discovery...")
+    graph = Graph()
+    edges = set()
+
+    # ##### Add all nodes and record colinear pairs as potential edges ##### #
+
+    logger.info("Adding nodes...")
+    for df in dict_of_dataframes.values():
+        for row in df.to_dict(orient="records"):
+            nodes = [(k, v) for k, v in row.items() if v != NA]
+            for c, n in nodes:
+                graph.add_node((c, n))
+                graph.add_edge((c, n), c)  # relate each node to its type
+            for combo in itertools.combinations(nodes, 2):
+                edges.add(combo)
+
+    counts = _get_node_counts(graph)
+    colwidth = len(max(counts.keys(), key=len))
+    count_block = [
+        f"\t{k:<{colwidth}} : {v if v > 0 else ''}" for k, v in counts.items()
+    ]
+    count_string = "\n".join(count_block)
+    logger.info(f"{sum(counts.values())} nodes added.\n{count_string}")
+
+    # ###### Add explicit (direct hierarchy) connections to the graph ###### #
+
+    logger.info("Adding direct edges...")
+
+    remaining_edges = set()
+    ignored_edges = set()
+
+    for a, b in edges:
+        if b[0] in HIERARCHY.nodes(a[0]):
+            graph.add_edge(a, b)
+        elif a[0] in HIERARCHY.nodes(b[0]):
+            graph.add_edge(b, a)
+        else:
+            if HIERARCHY.is_connected(a[0], b[0]):
+                remaining_edges.add((a, b))
+            elif HIERARCHY.is_connected(b[0], a[0]):
+                remaining_edges.add((b, a))
+            else:
+                ignored_edges.add((a, b))
+
+    logger.info(f"{_get_edge_counts(graph)} direct edges added.")
+    logger.info(f"{len(ignored_edges)} extra-hierarchical edges ignored.")
+
+    def prune_unneeded(edges):
+        # (A -> B) + (B -> C) doesn't also need (A -> C)
+        if edges:
+            logger.info(
+                f"{len(edges)} indirect edges left to insert BEFORE pruning."
+            )
+            graph.is_connected.cache_clear()
+            edges = [e for e in edges if not graph.is_connected(e[0], e[1])]
+            logger.info(
+                f"{len(edges)} indirect edges left to insert AFTER pruning."
+            )
+        return edges or []
+
+    edges = prune_unneeded(remaining_edges)
```

Comment:
needed because of https://github.com/kids-first/kf-lib-data-ingest/pull/493/files#diff-e190920f4698d285e201248a5c0b70b1R247-R250

Reply:
I think we should only do what is necessary for validation on direct connections because `prune_unneeded` is adding an extra couple minutes for largeish studies

Reply:
down to about a minute now

Reply:
I still don't see why you need to do this if you're only validating based on the direct connections? I understand you need it for checking gaps in the type hierarchy in the `gaps` method. But why do you need to check for gaps in the case where you're only evaluating direct connections? Without gap tests, validation executes in seconds vs minutes.

Reply:
I guess there's this question about what to do for data from the files that are not connected directly, then. We'll need a new report type for them.

Reply:
I think need to see an example of what the gaps analysis is reporting and how it could be useful because I'm not sure what action the user would take on reported implied connections? Is that supposed to be a signal to them to run the validation with include_implicit=True? Or is it a signal to them to that some entities are missing links? Because if it is the latter, I think you get that information from the relationship test errors right? 

For example, we have a file with particpant, specimen data, `participant-specimens.tsv`, and a file with participant, source gf data, `participant-files.tsv`. If I run validation with include_implicit=False, then I would expect to see relationship test errors telling me my files are missing specimens. What additional info would the gap analysis errors tell me in this case? That I should run validation with include_implicit=True?